openhcl/underhill_core/src/nvme_manager.rs

```diff
@@ -1,9 +1,62 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-//! Provides access to NVMe namespaces that are backed by the user-mode NVMe
-//! VFIO driver. Keeps track of all the NVMe drivers.
-
+//! Multi-threaded NVMe device manager for user-mode VFIO drivers.
+//!
+//! # Architecture Overview
+//!
+//! This module implements a multi-threaded actor-based architecture for managing NVMe devices:
+//!
+//! ```text
+//! NvmeManager (coordinator)
+//!   ├── NvmeManagerWorker (device registry via mesh RPC)
+//!   │   └── Arc<RwLock<HashMap<String, NvmeDriverManager>>> (device lookup)
+//!   │
+//!   └── Per-device: NvmeDriverManager
+//!       └── NvmeDriverManagerWorker (serialized per device via mesh RPC)
+//!           └── VfioNvmeDevice (wraps nvme_driver::NvmeDriver<VfioDevice>)
+//! ```
+//!
+//! # Key Objects
+//!
+//! - **`NvmeManager`**: Main coordinator, creates worker task and provides client interface
+//! - **`NvmeManagerWorker`**: Handles device registry, spawns tasks for concurrent operations  
+//! - **`NvmeDriverManager`**: Per-device manager with dedicated worker task for serialization
+//! - **`NvmeDriverManagerWorker`**: Serializes requests per device, handles driver lifecycle
+//! - **`VfioNvmeDevice`**: Implements `NvmeDevice` trait, wraps actual NVMe VFIO driver
+//! - **`VfioNvmeDriverSpawner`**: Implements `CreateNvmeDriver` trait for device creation
+//! - **`NvmeDiskResolver`**: Resource resolver for converting NVMe configs to resolved disks
+//! - **`NvmeDiskConfig`**: Configuration for NVMe disk resources (PCI ID + namespace ID)
+//!
+//! # Concurrency Model
+//!
+//! - **Cross-device operations**: Run concurrently via spawned tasks
+//! - **Same-device operations**: Serialized through per-device worker tasks
+//! - **Device registry**: Protected by `Arc<RwLock<HashMap<String, NvmeDriverManager>>>`
+//! - **Shutdown coordination**: `Arc<AtomicBool>` prevents new operations during shutdown
+//!
+//! # Lock Order (to prevent deadlocks)
```

Comment:
Maybe not in this PR, but could we _enforce_ this lock order via the typesystem? There's a talk that someone did at rustconf that I can't find the link to right now, where they had a networking stack with multiple locks and they enforced the ordering via the type system which removed a ton of bugs.

Reply:
could we track this as an issue maybe? is it possible to deadlock here if you do something wrong?